### PR TITLE
Allow fetching of historical data for CL-SIC

### PR DESCRIPTION
**Pull request by systemcatch**
- Handle old format xls files correctly.
- Read only the correct sheet from xls file.
- Add geothermal category.
- Avoid error if no generation in hour for fuel type.
- Add new plants.

Bruno, until they start updating the site again you can use target_datetime to get files after 6th November.

Output example

```shell
(EM-env) chris@ThinkPad:~/electricitymap$ python parsers/CL_SIC.py 
fetch_production(target_datetime=2016-01-01)
[{'zoneKey': 'CL-SIC', 'datetime': datetime.datetime(2016, 1, 2, 0, 0, tzinfo=tzfile('/usr/share/zoneinfo/Chile/Continental')), 'production': {'solar': 0.0, 'coal': 1888.2, 'wind': 966342.8999999998, 'biomass': 234.69999999999993, 'unknown': 0.7, 'hydro': 17361032.322899994, 'oil': 14.9, 'gas': 321.7}, 'storage': {}, 'source': 'sic.coordinador.cl'}, {'zoneKey': 'CL-SIC', 'datetime': datetime.datetime(2016, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Chile/Continental')), 'production': {'solar': 0.0, 'coal': 1903.4, 'wind': 1585315.8243000007, 'biomass': 244.70000000000002, 'unknown': 0.7, 'hydro': 14333647.57380001, 'oil': 15.1, 'gas': 285.5}, 'storage': {}, 'source': 'sic.coordinador.cl'}, {'zoneKey': 'CL-SIC', 'datetime': datetime.datetime(2016, 1, 1, 2, 0, tzinfo=tzfile('/usr/share/zoneinfo/Chile/Continental')), 'production': {'solar': 0.0, 'coal': 1904.4, 'wind': 1185011.3194, 'biomass': 234.70000000000002, 'unknown': 0.7, 'hydro': 13837277.098700004, 'oil': 15.2, 'gas': 268.8}, 'storage': {}, 'source': 'sic.coordinador.cl'}, {'zoneKey': 'CL-SIC', 'datetime': datetime.datetime(2016, 1, 1, 3, 0, tzinfo=tzfile('/usr/share/zoneinfo/Chile/Continental')), 'production': {'solar': 0.0, 'coal': 1896.2, 'wind': 729369.0264000002, 'biomass': 240.90000000000003, 'unknown': 0.7, 'hydro': 12849447.199200004, 'oil': 13.7, 'gas': 271.0}, 'storage': {}, 'source': 'sic.coordinador.cl'},...........snipped 
```

ref #1683 #1257

### CODE REVIEW

File: parsers/CL_SIC.py [hunk 1]
```diff
@@ -210,36 +211,66 @@
...
-    date_no_tz = arrow.get(date_str, "YYMMDD")
-    date = date_no_tz.replace(tzinfo='Chile/Continental')
-
+    if not target_datetime:
+        s = session or requests.Session()
+        document_url = 'https://sic.coordinador.cl/informes-y-documentos/fichas/operacion-real/'
+        req = s.get(document_url)
+        soup = BeautifulSoup(req.text, 'html.parser')
+
+        # Find the latest file.
```
Comment (brunolajoie) [inline parsers/CL_SIC.py]: In the eventuality that their website is not maintained anymore and the html <a> tag is never udpated, one way to find the latest xls file is to call directly the excel URL "coordinador.cl/wp-content/uploads/estadisticas/operdiar/18/OP181127.xls" with the latest date returning a "status 200" request starting from date = now
One other way to find

Comment (brunolajoie) [reply parsers/CL_SIC.py]: That's more a nice-to-have than a must have, but, i'm afraid that SIC is not maintaining this webpage anymore in favor of a newest one

File: parsers/CL_SIC.py [hunk 2]
```diff
@@ -210,36 +211,66 @@
...
+        extension = generation_link["href"]
+        base_url = "https://sic.coordinador.cl"
+        data_url = base_url + extension
+
+        date_pattern = r'OP(\d+)\.xls'
+        date_str = re.search(date_pattern, extension).group(1)
+        date_no_tz = arrow.get(date_str, "YYMMDD")
+        date = date_no_tz.replace(tzinfo='Chile/Continental')
+    else:
+        lookup_date = target_datetime.format('YYMMDD')
```
Comment (brunolajoie) [inline parsers/CL_SIC.py]:
```suggestion
        target_datetime = arrow.get(target_datetime)
        lookup_date = target_datetime.format('YYMMDD')
```
That should make the historical fetching option compatible with test_parser.py & our backend

File: parsers/CL_SIC.py [hunk 3]
```diff
@@ -276,11 +307,17 @@ def thermal_processer(df, logger):
...
     unknown_generation = []
 
-    for plant in thermal_plants.keys():
-        plant_vals = thermal_df.loc[plant].to_dict()
-        plant_type = thermal_plants[plant]
+    for plant in THERMAL_PLANTS.keys():
+        try:
+            plant_vals = thermal_df.loc[plant].to_dict()
+        except KeyError:
+            # plant is missing from df
```
[the remark this answers is not in the document]
Comment (systemcatch) [reply parsers/CL_SIC.py]: Might be better to loop over the plants in the df rather than the dict mapping here.

File: parsers/CL_SIC.py (hunk 1 above)
Comment (systemcatch) [reply parsers/CL_SIC.py]: Yes right now you can use target datetime to get files after 6th November, we can change this if the website stops updating permanently.

File: parsers/CL_SIC.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (systemcatch) [reply parsers/CL_SIC.py]: Hopefully my changes are ok, all unmapped plants are being sent to the logger.

File: parsers/CL_SIC.py [hunk 4]
```diff
@@ -276,11 +308,13 @@ def thermal_processer(df, logger):
...
     biomass_generation = []
+    geothermal_generation = []
     unknown_generation = []
 
-    for plant in thermal_plants.keys():
+    for plant in data_plants:
         plant_vals = thermal_df.loc[plant].to_dict()
-        plant_type = thermal_plants[plant]
+
+        plant_type = THERMAL_PLANTS.get(plant, 'unknown')
```
Comment (corradio) [inline parsers/CL_SIC.py]: sorry to be annoying, but should we log the case where a `plant_type` is not the `THERMAL_PLANTS` in order to add it in the future?

Comment (systemcatch) [reply parsers/CL_SIC.py]: L305 should catch that.

Comment (corradio) [reply parsers/CL_SIC.py]: ha! didn't see it. 👍

**Comment by systemcatch** [comment]
Just an FYI guys, arrow has some weird results when parsing certain strings. Good to know for any target_datetime implementation.

https://github.com/crsmithdev/arrow/issues/519
https://github.com/crsmithdev/arrow/issues/91

**Pull request merged**